Everything

michalrud opened the pull request:
Review of all the code written.

Commit:
+CO2, Humidity

Commit:
Handler tokens (to gracefully unsubscribe handlers in case of service disabling)

Commit:
Catch -> GTest

Mocks are useful, and rolling own framework is an overkill

Commit:
gitlab ci

Commit:
Basic Arduino mock, more Wire mock stuff

Commit:
I2CHelpers

Commit:
ThetaCloudLight rewritten to I2CUtils
Float to String util
MockThetaCloud

Commit:
Humidity to I2C Helpers
Fixed a bug where Temperature reported Humidity instead

Commit:
Fixed compilation problems

Commit:
ThetaCloudCO2 to I2C Helpers

Commit:
Initial Sphinx documentation/webpage

Commit:
I2CHelpers API documentation + Initial documentation stuff

Commit:
little cleanup

Commit:
Fixed gitlabci mermaid installation

Commit:
Part of ThetaCloud base documentation

Commit:
Sensor* -> Device*

Some devices (eg. relays) are not technically sensors.

Commit:
Added test which assures that nothing happens when Emit callback is not added

Commit:
Added proer handling of multiple subscriptions of WriteHandlers

Commit:
More ThetaCloud base documentation

Commit:
Added the ThetaCloud::emit() method

Commit:
Rest of ThetaCloud base API documentation

Commit:
Added Arduino example and description

Commit:
Updated README

Commit:
Simple ThetaCloud programming/flashing tutorial

Commit:
renamed floatToString -> to_string

Commit:
ThetaCloudI2CScanner

Commit:
Fixed tests compilation

Commit:
Introduced the "consumption" term in docs

Commit:
Renamed: ThetaCloudCO2 -> ThetaCloudEnvironment

Commit:
Fixed reStructuredText formatting typo

Commit:
Added referenced libraries

Commit:
Added the optional support for MPL3115A2

Commit:
ThetaCloudRelay documentation

Commit:
ThetaCloudLight documentation

Commit:
Forced color output in GTest.

Commit:
ThetaCloudHumidity documentation

Commit:
Added units into documentation

Commit:
Renamed "Programming" to "Flashing" to avoid ambiguity

Commit:
Library usage tutorial

Commit:
Field Lists -> Sphinx Domains

Commit:
Sphinx formatting improvements

Commit:
Added const strings to ThetaCloudRelay

Commit:
Library usage reference part 4

Commit:
Small improvements

Commit:
ThetaCloudRelay::RELAY{1,2} made const char* to make using it easier

Commit:
Made mocks stricter

Commit:
ThetaCloudPIR

Commit:
Removed unused "enabled" variables

Commit:
enabled CTest

Commit:
Added coverage, -Werror, -Wall

Commit:
Added forgotten relay tests

Commit:
show test output in case of failure

Commit:
Test cleanup

Commit:
Formatting

Commit:
Added formatting checking in gitlabci

Commit:
Removed installing of python3-pip from formatting job

as it's not necessary there

Commit:
Removed the need for mermaid in documentation

Commit:
Added line numbers to documentation

Commit:
First part of "implementing support for new boards"

Commit:
Clarified, that emit may be called multiple times as well as not called at all

Commit:
Documentation on registering write handlers

Commit:
First commit with code

Commit:
removed mistakenly added Makefile

wrazik commented on CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+include_directories("${source_dir}/googletest/include"
+                    "${source_dir}/googlemock/include")
+
+set(CMAKE_CXX_FLAGS "-g -O0 -Wall -Werror -fprofile-arcs -ftest-coverage")
```
Please consider extra compilation flags, like: https://github.com/lefticus/cppbestpractices/blob/master/02-Use_the_Tools_Available.md#gcc--clang

wrazik commented on CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+target_link_libraries( lightTests libgtest libgmock gcov )
+add_test(NAME light COMMAND lightTests )
+
+set (humidity_tests test/main.cpp test/humidity_tests.cpp)
```
Do you need separate binaries for all of test cases?

michalrud replied:
Not really. Is having tests in separate libraries bad?

The only reason I had was that each of these modules should be independent of each other, this way we can be sure that they can work without other modules

wrazik commented on DeviceHandlerToken.h:
```diff
@@ -0,0 +1,30 @@
+    {
+    }
+
+    const std::function<void()> deleter;
```
Maybe instead of std::function you can use function pointer?
Like:
```
using FunctionPtr = void(*)();
const FunctionPtr deleter;
```
std::function adds bloat, function ptr is faster, I've created a benchmark to show it: http://quick-bench.com/eDyrxB6vGZWQ0IUrPMTSkMSxSsA

michalrud replied:
I've used `std::function` as I am passing lambdas there. I think I should store these lambdas somewhere as the pointers would get invalid after the lambda would get out of scope, so it would complicate the code quite much - or am I missing something?

wrazik commented on I2CHelpers.h:
```diff
@@ -0,0 +1,54 @@
+#include <array>
+#include <utility>
+
+const unsigned int READ_RETRIES = 10;
```
can be conxtexpr

michalrud replied:
:heavy_check_mark:

wrazik commented on ThetaCloud.cpp:
```diff
@@ -0,0 +1,64 @@
+#include "Wire.h"
+#include <cassert>
+
+const int TC_I2C_SDA = 2;
```
constexpr?

michalrud replied:
:heavy_check_mark:

wrazik commented on ThetaCloud.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <cassert>
+
+const int TC_I2C_SDA = 2;
+const int TC_I2C_SCL = 14;
```
constexpr?

michalrud replied:
:heavy_check_mark:

wrazik commented on ThetaCloud.cpp:
```diff
@@ -0,0 +1,64 @@
+        handler(this->dataCallback);
+}
+
+ThetaCloud thetaCloud;
```
Do we need global variable here? Maybe you should consider to remove it?

michalrud replied:
As this library is targeted to beginner Arduino programmers, I've tried to adjust the code to how other libraries worked to make it easier for people to start working with it - that's how `Wire` or `Serial` standard libraries work, for example.

Though after looking at some other libraries, it looks like people randomly do the static instance thing or not, so I will go ahead and remove this.

wrazik commented on ThetaCloud.h:
```diff
@@ -0,0 +1,45 @@
+        initialized(false)
+    {
+    }
+    ~ThetaCloud()
```
maybe = default?

michalrud replied:
:heavy_check_mark:

wrazik commented on ThetaCloud.h:
```diff
@@ -0,0 +1,45 @@
+
+    Emit dataCallback;
+    bool initialized;
+    std::list<DeviceReadHandler> deviceReadHandlers;
```
why std::list not std::vector?

michalrud replied:
no reason, I will change it as vectors would fit more nicely

michalrud replied:
:heavy_check_mark:

michalrud replied:
Turns out this was a wrong decision - I forgot that the `list` was used as iterators are not invalidated when items are added or removed from list, what is not the case in `vector`. Fortunately the tests found out pretty quickly ;)

wrazik commented on ThetaCloudEnvironment.cpp:
```diff
@@ -0,0 +1,68 @@
+const uint8_t MICS_VZ_89_ADDRESS = 0x70;
+const uint8_t GET_STATUS_COMMAND = 0b00001001;
+
+ThetaCloudEnvironment::ThetaCloudEnvironment()
```
=default?

michalrud replied:
:heavy_check_mark:

Commit:
trying out travis

Commit:
uncrustify -> clang-format

Commit:
Added additional compiler flags

Commit:
Fixed issues found by adding additional flags

Commit:
removed flags not supported on CI

Commit:
const -> constexpr

Commit:
list -> vector

Commit:
removed empty constructors, destructors

Commit:
forgotten const -> constexpr, ctest output on failure

Commit:
valgrinded tests by default

Commit:
Changing list -> vector was a bad change, reverted.

Issue was - in list iterators are not invalidated
when items are added or deleted, while in vector
they are. Handler tokens depended on iterators staying
valid despite containers contents changing.

Commit:
valgrind tests can be disabled with `ctest -E Valgrind`

Commit:
Removed the method signatures, as it's not really required

Commit:
Removed static initializations

Commit:
fixed compilation on Arduino and examples
